### DISCUSSION TITLE
HTTP Post Method not working when using nitro without nuxt

### DISCUSSION THREAD

**Discussion by folamy** [nitrojs/nitro#1677] [Q&A]
### Environment

Linux OS
Ubuntu 22.04

### Reproduction

[reproduction](https://stackblitz.com/edit/github-u6vrxc?file=routes%2Fconfirm-code.post.ts)

### Describe the bug

I set up a nitro server mainly as an API/backend for a mobile app
everything seems to work fine but following the examples on the [documentation](https://nitro.unjs.io/guide/routing#specific-request-method) , HTTP POST requests return no response from the server:
Adding .post.ts to my route/filename
![image](https://github.com/unjs/nitro/assets/4560948/adf4916c-ace0-4641-aeb9-68fc981ea7ac)
This does nothing!
And when I remove the .post suffix, h3 composable utils like readBody() throws error
H3Error: HTTP method is not allowed
![image](https://github.com/unjs/nitro/assets/4560948/fc1afe7c-a5d2-42ab-a19c-0cdfa197e2ba)

What am I doing wrong?

**Comment by pi0** [comment]
Would you please provide a reproduction? 🙏🏼 (please make sure you are actually making a POST request)

**Comment by folamy** [comment]
> > Would you please provide a reproduction? 🙏🏼 (please make sure you are actually making a POST request)
> 
@pi0 
> https://stackblitz.com/edit/github-u6vrxc?file=routes%2Fconfirm-code.post.ts

**Comment by folamy** [comment]
I just used HTTP client from vscode and Post method works just as expected but when using ofetch or axios from my frontend project, I get the same error

**Comment by folamy** [comment]
I am making the post request from a different project also running local server @port:8080
![image](https://github.com/unjs/nitro/assets/4560948/3974fc26-4bfe-4cb2-a3cf-d7d121e3085f)

**Comment by pi0** [comment]
I have forked your reproduction with Nitro 2.6. Fetch body goes as expected. Please check this if I have missed something:

https://stackblitz.com/edit/github-u6vrxc-kwmdpb?file=routes%2Findex.ts,routes%2Fconfirm-code.post.ts

<img width="1080" alt="image" src="https://github.com/unjs/nitro/assets/5158436/29d03851-650d-4e5b-8983-34d38b3babbd">


---

Can you check having latest nitro (and fresh deps with h3 1.8?). It might be also another error hidden because of cors issues. If you can share a screenshot of your devtools network tab from external project that is making HTTP call might be helpful to investigate.

**Comment by folamy** [comment]
I believe this is an internal error, I was able to solve this with a one on one help from @danielroe  by creating a file in the middleware folder like this
```js
export default defineEventHandler((event) => {
  if (event.node.req.method === 'OPTIONS')
    return null
})

**Comment by danielroe** [comment]
To clarify, this should be in the same handler (which should not be only a POST handler) rather than in the middleware directory.

The issue was that we had to handle CORS requests in this particular setup

**Comment by folamy** [comment]
> To clarify, this should be in the same handler (which should not be only a POST handler) rather than in the middleware directory.
> 
> The issue was that we had to handle CORS requests in this particular setup

@danielroe 
Doing that in the same handler will terminate code? Can you show me an example sir, pls?

**Comment by danielroe** [comment]
```ts
export default defineEventHandler(async (event) => {
  if (event.method === 'OPTIONS') {
    // set your options here
    handleCors(event, {})
    return null
  }

  assertMethod(event, ['POST'])
  const body = await readBody(event)
  console.log('Body:', body)
  // TODO: Handle body and add user
  return { body }
})
```

**Reply by folamy** [reply]
@danielroe, this works if only I remove the .post suffix from the file and comment out the handleCors() composable

**Comment by folamy** [comment]
> ```ts
> export default defineEventHandler(async (event) => {
>   if (event.method === 'OPTIONS') {
>     // set your options here
>     handleCors(event, {})
>     return null
>   }
> 
>   assertMethod(event, ['POST'])
>   const body = await readBody(event)
>   console.log('Body:', body)
>   // TODO: Handle body and add user
>   return { body }
> })
> ```

@danielroe,
if this condition is true (event.method === 'OPTIONS') and you return null in the api route file.ts, it will ternminate every other code from running. believe no codes runs after a return statement in javascript.

**Comment by danielroe** [comment]
I'm not sure I understand the issue. The OPTIONS request should not perform additional work other than returning the correct headers.

**Comment by folamy** [comment]
> ```ts
> export default defineEventHandler(async (event) => {
>   if (event.method === 'OPTIONS') {
>     // set your options here
>     handleCors(event, {})
>     return null
>   }
> 
>   assertMethod(event, ['POST'])
>   const body = await readBody(event)
>   console.log('Body:', body)
>   // TODO: Handle body and add user
>   return { body }
> })
> ```
@danielroe, this only works if I remove the .post suffix from the file and comment out the handleCors() composable.
